v2 bringup

TDHolmes opened the pull request:
Bring up of the second version of the board

  - [x] build system refactor
  - [x] basic V2 functionality
  - [x] new sensor support

TDHolmes committed:
some initial rework of main

TDHolmes committed:
some initial rework of main

TDHolmes committed:
adding final v2 schematic

TDHolmes committed:
adding final v2 schematic

TDHolmes committed:
cleaning up some unused calls for v2

TDHolmes committed:
cleaning up some unused calls for v2

TDHolmes committed:
add v1 schematic

TDHolmes committed:
add v1 schematic

TDHolmes committed:
v2 oscillator working I believe?

TDHolmes committed:
v2 oscillator working I believe?

TDHolmes committed:
reconfigure I2C and other peripherals for v2

TDHolmes committed:
reconfigure I2C and other peripherals for v2

TDHolmes committed:
point to the correct .elf in the build size and usage subcommands

TDHolmes committed:
point to the correct .elf in the build size and usage subcommands

TDHolmes committed:
cleanup and fix

TDHolmes committed:
cleanup and fix

TDHolmes committed:
initialize UART and cleanup some LSM stuff

TDHolmes committed:
initialize UART and cleanup some LSM stuff

TDHolmes committed:
fix clocking issue

TDHolmes committed:
fix clocking issue

TDHolmes committed:
add basic scheduler

TDHolmes committed:
add basic scheduler

TDHolmes committed:
USB support!?!

TDHolmes committed:
USB support!?!

TDHolmes committed:
small changes

TDHolmes committed:
small changes

TDHolmes committed:
USB datapair protection datasheet. Definitely should've used this

TDHolmes committed:
USB datapair protection datasheet. Definitely should've used this

TDHolmes committed:
Adding some communication arch documentation

TDHolmes committed:
v2 pictures

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -77,6 +94,33 @@ void clear_critical_errors(void) {
 }
 
+extern __IO uint8_t Receive_Buffer[64];
+extern __IO uint32_t Receive_length;
```
There seems to be a lot of inconsistencies with variable naming standards, such as `Aaaa_Aaaa`, `Aaaa_aaaa`, `aaaa_aaaa`. Worth looking through the code and trying to standardize to make everything more cohesive

TDHolmes replied:
Unfortunately due to my standard clashing with STM’S HAL naming convention. Definitely will clean up the main code areas

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
+        if (bDeviceState == CONFIGURED) {
+            CDC_Receive_DATA();
+            /*Check to see if we have data yet */
+            if (Receive_length  != 0) {
```
nit: Double space after `Receive_length`

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
+            /*Check to see if we have data yet */
+            if (Receive_length  != 0) {
+                if (packet_sent == 1)
+                    CDC_Send_DATA ((unsigned char*)Receive_Buffer,Receive_length);
```
Recommend adding curly brackets, again to help reduce possible confusion in the future make the code as robust as possible against misunderstanding scope/flow

TDHolmes replied:
Def

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
 {
-    ret_t retval;
-    uint32_t subcount = 0;
+    uint8_t i;
```
consider giving this variable a more descriptive name since it is not clear what it is used for until you dig into `scheduler_add`

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
+    UART_sendString("Initialized\r\n");
 
+    uint32_t sleep_until = 0;
+    while (true) {
```
You are using `while (true)` and `while (1)` both within this same function, would be better to pick one format or the other simply for consistency

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
 
+    UART_sendString("Initialized\r\n");
 
+    uint32_t sleep_until = 0;
```
I personally think it is a good idea to follow the `ANSI C89` standard and declare all local variables at the beginning of the scope they are used in, but will leave this discretion up to you

TDHolmes replied:
Main is real hacky right now haha

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
+        sleep_until += time_until_next_cb_ms;
+
+        if (time_until_next_cb_ms > 0) {
+            UART_sendString("Sleeping for "); UART_sendint(time_until_next_cb_ms); UART_sendString(" ms\r\n");
```
Any specific reason to not split these up to individual lines? I git the jist of what you are going for here but personally don't think it is very readable

TDHolmes replied:
This is all just debug code that will be deleted before this goes in

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
     }
-    #endif
+    tick += 1;
+    return RET_OK;
```
Any point in this function having a return value if it can only ever return `RET_OK`?

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
-    } else {
-        second_count++;
+    // toggle LED every 10,000 itterations over the workloop
+    if (i++ >= 10000) {
```
Same comment as made above, better to be verbose and easy to understand rather than have something that could be easily misunderstood down the road

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
+    return RET_OK;
+}
+
+
```
nit: extra new line

BroderickCarlin commented on Pensel/firmware/common.h:
```diff
@@ -42,6 +42,10 @@
     #define min(v1, v2) (v1 < v2 ? v1 : v2)
 #endif
 
+#if !defined(NULL)
```
`#ifndef NULL`

BroderickCarlin commented on Pensel/firmware/common.h:
```diff
@@ -42,6 +42,10 @@
 #endif
 
+#if !defined(NULL)
+    #define NULL ((void *) 0)
```
Any reason for defining this yourself and not using `cstddef.h`?

TDHolmes replied:
Didn’t know about this!

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
-        wdg_pet();
+/* --- Some common callbacks to be run --- */
+
+ret_t heartbeat(int32_t * new_callback_time_ms) {
```
You jump between having the function curly brace on the same line as the declaration and on the following line, just a heads up for consistency

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -0,0 +1,118 @@
+ret_t scheduler_init(schedule_t * schedule)
+{
+    schedule->num_slots_used = 0;
+    for (uint8_t i=0; i < NUM_SLOTS; i++) {
```
nit: spaces around `=`

TDHolmes replied:
I’ve gone back and forth on this style. I think I might like what is there from python coding

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -0,0 +1,118 @@
+                    ret_t (*callback)(int32_t * callback_time_ms),
+                    uint8_t * schedule_id)
+{
+    bool slot_found = false;
```
You know the capacity of your schedule array, and the number of filled slots, so is this variable really necessary when you could just stick with the check you have on line 36?

TDHolmes replied:
Yeah I suppose the if check around that var will never be reached!

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -0,0 +1,118 @@
+            }
+
+            // Check if this will be the next callback we need to run next time
+            if (time_until_next_run  < next_cb_time_ms) {
```
nit: extra white space after `time_until_next_run`

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.h:
```diff
@@ -0,0 +1,32 @@
+    ret_t (*callbacks[NUM_SLOTS])(int32_t * callback_time_ms);
+    int32_t callback_times_ms[NUM_SLOTS];
+    int32_t callback_last_run_ms[NUM_SLOTS];
+} schedule_t;
```
Thoughts on creating a single `callback_t` struct that will hold the callback and time stamps? This struct could then simply the last three elements to a single `callback_t callbacks[NUM_SLOTS]` entry.

TDHolmes replied:
I like it!

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -0,0 +1,118 @@
+    }
+    return RET_OK;
+}
+
```
In some of your source files you have 1 new line after functions, while in this one you have 2. Just a reminder that consistency is good ✌️

TDHolmes replied:
Fact

TDHolmes replied:
Maybe I should make a python linter...

TDHolmes replied:
clang-format looks promising. Spinning this off into it's own branch for later

BroderickCarlin commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -0,0 +1,118 @@
+        return RET_LEN_ERR;
+    }
+    // TODO: disable timer interrupt here
+    for (uint8_t i=0; i < NUM_SLOTS; i++) {
```
nit: spaces around `=`. If this is a personal style preference ignore

BroderickCarlin commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -77,6 +94,33 @@ void clear_critical_errors(void) {
+    if (ret != HAL_OK) {
+        check_retval_fatal(__FILE__, __LINE__, (ret_t)ret);
+    }
+
```
nit: extra newline

TDHolmes committed:
Some USB code cleanup

TDHolmes committed:
Some USB code cleanup

TDHolmes committed:
Initial creation of LSM9DS1 driver code

TDHolmes committed:
Initial creation of LSM9DS1 driver code

TDHolmes committed:
Reformatting descriptor comment structure. No real code change

TDHolmes committed:
Reformatting descriptor comment structure. No real code change

TDHolmes committed:
Some progress in making LSM9DS1 driver

TDHolmes committed:
Some progress in making LSM9DS1 driver

TDHolmes committed:
Some more cleanup

TDHolmes committed:
Some more cleanup

TDHolmes commented on Pensel/firmware/applications/pensel_v2.c:
```diff
@@ -86,108 +130,125 @@ void clear_critical_errors(void) {
-        // let the user know roughly how many workloops are ocurring
-        if (subcount == 100000) {
+        // toggle LED every 10,000 itterations over the workloop
+        if (ledi++ >= 10000) {
```
Agreed!

TDHolmes committed:
Attempts to get clang working

TDHolmes committed:
Attempts to get clang working

TDHolmes committed:
Many small refactors; USB serial works!

TDHolmes committed:
Many small refactors; USB serial works!

TDHolmes committed:
verbose building

TDHolmes committed:
verbose building

TDHolmes committed:
refactor some usb; break CDC stuff into it's own module

TDHolmes committed:
refactor some usb; break CDC stuff into it's own module

TDHolmes committed:
re-enable the scheulder stuffz

TDHolmes committed:
re-enable the scheulder stuffz

TDHolmes committed:
refactor

TDHolmes committed:
refactor

TDHolmes committed:
cleanup main a bit

TDHolmes committed:
cleanup main a bit

TDHolmes committed:
Continued refactor/cleanup

TDHolmes committed:
Continued refactor/cleanup

TDHolmes committed:
initialize state, use correct error handler function

TDHolmes committed:
initialize state, use correct error handler function

TDHolmes committed:
Death to v1 code

TDHolmes committed:
Death to v1 code

TDHolmes committed:
extra space

TDHolmes committed:
extra space

TDHolmes committed:
cammelCase snake case fixup

TDHolmes committed:
cammelCase snake case fixup

TDHolmes committed:
Fix I2C init stuff

TDHolmes committed:
Fix I2C init stuff

TDHolmes committed:
LSM9DS1 is talkin!

TDHolmes committed:
LSM9DS1 is talkin!

TDHolmes committed:
fix build script default arg to be valid

TDHolmes committed:
fix build script default arg to be valid

TDHolmes commented on Pensel/firmware/modules/utilities/scheduler.c:
```diff
@@ -108,7 +108,7 @@ int32_t scheduler_run(schedule_t * schedule, uint32_t current_time_ms)
 
             // Check if this will be the next callback we need to run next time
-            if (time_until_next_run  < next_cb_time_ms) {
+            if (time_until_next_run < next_cb_time_ms) {
```
I commit early, and commit often :P I plan on squashing to a few commits with your recommended git commands at the end though! And the end is coming soon :p

TDHolmes committed:
Merge branch 'tholmes/v2-bringup' of github.com:TDHolmes/Harma into tholmes/v2-bringup

TDHolmes committed:
LSM9DS1 inits and outputs data, interrupt-driven.

Also, some CDC serial improvements

TDHolmes committed:
add logging module

TDHolmes committed:
Many changes:

- Doxygen updates
- Docstring updates (everywhere)
- USB HID files
- Queue up gyro/mag/accel packets
- Accessor methods for those packets
- Control reports to get said packets
- Updated interrupt prioritizations
- Auto select files/ports if they are the only choice in gdb startup 
script

TDHolmes committed:
clang format fixes

TDHolmes committed:
more clang-format changes; delete defunct UART reports

TDHolmes commented:
V2 support is pretty much done - gonna merge this in to mainline!
